File: app/lib/models/Species.dart
import 'ConservationStatus.dart';
import 'Habitat.dart';
import 'taxonomy/Genus.dart';

class Species {
  final int id;
  final String scientificName;
  final String commonName;
  final ConservationStatus? conservationStatus;
  final List<Habitat> habitats;
  final Genus? genus;
  final String? imageUrl;

  Species({
    required this.id,
    required this.scientificName,
    required this.commonName,
    this.conservationStatus,
    required this.habitats,
    this.genus,
    this.imageUrl,
  });

  factory Species.fromJson(Map<String, dynamic> json) {
    var habitatsList = (json['habitats'] as List)
        .map((habitatJson) => Habitat.fromJson(habitatJson))
        .toList();
    ConservationStatus? conservationStatus;
    if (json['conservationStatus'] != null) {
      conservationStatus =
          ConservationStatus.fromJson(json['conservationStatus']);
    }
    Genus? genus;
    if (json['genus'] != null) {
      genus = Genus.fromJson(json['genus']);
    }
    return Species(
      id: json['id'],
      scientificName: json['scientific_name'],
      commonName: json['common_name'],
      imageUrl: json['image_url'],
      conservationStatus: conservationStatus,
      habitats: habitatsList,
      genus: genus,
    );
  }
}

File: api/routes/species.js
const express = require("express");
const {
  getAllSpecies,
  getSingleSpecies,
  createSpecies,
  updateSpecies,
  deleteSpecies,
} = require("../controllers/species");

const { models } = require("../models");
const advancedResults = require("../middleware/advancedResults");

const router = express.Router();

// Public routes
router
  .route("/")
  .get(
    advancedResults(models.Species, {
      include: [
        {
          model: models.Genus,
          as: "genus",
          attributes: ["name"],
          include: [
            {
              model: models.Family,
              as: "family",
              attributes: ["name"],
              include: [
                {
                  model: models.Order,
                  as: "order",
                  attributes: ["name"],
                  include: [
                    {
                      model: models.Class,
                      as: "class",
                      attributes: ["name"],
                      include: [
                        {
                          model: models.Phylum,
                          as: "phylum",
                          attributes: ["name"],
                          include: [{ model: models.Kingdom, as: "kingdom", attributes: ["name"] }],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
        {
          model: models.ConservationStatus,
          as: "conservationStatus",
          attributes: ["description", "severity", "name"],
        },
        {
          model: models.Characteristic,
          as: "characteristics",
          attributes: ["id", "characteristic_type", "value", "units"],
        },
        {
          model: models.Habitat,
          as: "habitats",
          through: { attributes: [] }, // Ẩn bảng trung gian
          attributes: ["id", "name", "description", "climate", "temperature", "humidity"],
        },
      ],
      attributes: ["id", "scientific_name", "common_name", "taxonomy_browser", "image_url", "createdAt"],
    }),
    getAllSpecies,
  ) // Get all species
  .post(createSpecies); // Create new species (Private)

router
  .route("/:id")
  .get(getSingleSpecies) // Get single species
  .put(updateSpecies) // Update species (Private)
  .delete(deleteSpecies); // Delete species (Private)

module.exports = router;
